src/utilities/buildASTSchema.js

```diff
@@ -518,7 +526,8 @@ export function getDescription(node: { loc?: Location }): ?string {
  * document.
  */
 export function buildSchema(source: string | Source): GraphQLSchema {
-  return buildASTSchema(parse(source));
+  const sourceObj = typeof source === 'string' ? new Source(source) : source;
```

Comment:
> In order to avoid recreating the Source object many times in invariantError, buildASTSchema now does so explicitly. But is this necessary?

Maybe we could construct a custom function globally to do this. I don't have a good answer for this as well, but yeah, passing an argument just for `invariantError` (or whatever we'll end up with) seems a bit unnecessary.

Reply:
Also note that ast nodes include a reference to the source they come from, so you shouldn't need to pass the second argument around